File: Difficulty: Medium/Unique Paths in a Grid/unique-paths-in-a-grid.cpp
class Solution {
  public:
    int recurSol(vector<vector<int>>& grid, int m, int n) {
        
        if(m<0 || n<0 || grid[m][n]==1) return 0;
        if(m==0 && n==0) return 1;
        int x = recurSol(grid,m-1,n);
        int y = recurSol(grid,m,n-1);
        return x+y;
    }
    int memoSol(vector<vector<int>>& grid, int m, int n, vector<vector<int>>& dp) {
        if(m<0 || n<0 || grid[m][n]==1) return 0;
        
        if(dp[m][n]!=-1) return dp[m][n];
        
        if(m==0 && n==0) return dp[m][n] = 1;
        
        
        
        int x = memoSol(grid,m-1,n,dp);
        int y = memoSol(grid,m,n-1,dp);
        return dp[m][n] = x+y;
    }
    int uniquePaths(vector<vector<int>> &grid) {
        // code here
        int m = grid.size();
        int n = grid[0].size();
        
        // return recurSol(grid,m-1,n-1);
        vector<vector<int>> dp(m,vector<int>(n,-1));
        return memoSol(grid,m-1,n-1,dp);

    }
};

File: Difficulty: Medium/Unique Paths in a Grid/unique-paths-in-a-grid.java
class Solution {
    public int solve(int n, int m, int[][] grid) {
        if(n>=0 && m>=0 && grid[n][m]==1) return 0;
        if(n==0 && m==0) return 1;
        if(n<0 || m<0) return 0;
        
        
        int up = solve(n-1,m,grid);
        int left = solve(n,m-1,grid);
        
        return (up+left);
        
        
    }
    
    public int solve2(int n, int m, int[][] grid, int[][] dp) {
        
        if(n>=0 && m>=0 && dp[n][m]!=-1) return dp[n][m];
        
        if(n>=0 && m>=0 && grid[n][m]==1) {
            dp[n][m] = 0;
            return 0;
        }
        if(n==0 && m==0) return 1;
        if(n<0 || m<0) return 0;
        
        
        
        int up = solve2(n-1,m,grid,dp);
        int left = solve2(n,m-1,grid,dp);
        
        return dp[n][m] = (up+left);
        
    }
    
    public int uniquePaths(int[][] grid) {
        // code here
        int n = grid.length;
        int m  = grid[0].length;
        
        int ans = 0;
        
        int[][] dp = new int[n][m];
        
        for(int i =0;i<n;i++) Arrays.fill(dp[i],-1);
        
        ans = solve2(n-1,m-1,grid,dp);
        
        return ans;
    }
};